// @alicloud/mp-db-service/dist/umd/mongo/model/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const codec_1 = require("../codec");
const constant_1 = require("../constant");
class BaseCommand {
    constructor(argMap) {
        this._schema = {};
        this._argMap = {};
        this._argMap = argMap;
        if (!this._argMap.options) {
            this._argMap.options = {};
        }
    }
    get permission() {
        switch (this._permission) {
            case constant_1.PERMISSION.CREATE:
            case constant_1.PERMISSION.UPDATE:
            case constant_1.PERMISSION.DELETE:
                return constant_1.PERMISSION.WRITE;
            default:
                return this._permission;
        }
    }
    get argMap() {
        return this._argMap;
    }
    get schema() {
        return this._schema;
    }
    augmentOptions(options) {
        this.argMap.options = Object.assign({}, this.argMap.options, options);
    }
    inspect() {
        const args = clean(JSON.parse(JSON.stringify(this.argMap)));
        function clean(d) {
            for (const k of Object.keys(d)) {
                if (codec_1.isObject(d[k])) {
                    if (Object.keys(d[k]).length === 0) {
                        delete d[k];
                    }
                    else {
                        clean(d[k]);
                    }
                }
                if (d[k] === undefined) {
                    delete d[k];
                }
            }
            return d;
        }
        return Object.assign(Object.assign({}, args), { command: this.name });
    }
}
exports.BaseCommand = BaseCommand;
class AggregateCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'aggregate';
        this._schema = {
            pipeline: 'array',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.AGGREGATE;
    }
}
exports.AggregateCommand = AggregateCommand;
class CountCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'count';
        this._schema = {
            query: 'object',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.READ;
    }
}
exports.CountCommand = CountCommand;
class DistinctCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'distinct';
        this._schema = {
            key: 'string',
            query: 'object',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.READ;
    }
}
exports.DistinctCommand = DistinctCommand;
class FindDocumentCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'findOne';
        this._schema = {
            query: 'object',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.READ;
    }
}
exports.FindDocumentCommand = FindDocumentCommand;
class FindDocumentsCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'find';
        this._schema = {
            query: {
                type: 'object',
                required: false,
            },
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.READ;
    }
}
exports.FindDocumentsCommand = FindDocumentsCommand;
class InsertDocumentCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'insertOne';
        this._schema = {
            doc: 'field',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.CREATE;
    }
}
exports.InsertDocumentCommand = InsertDocumentCommand;
class InsertDocumentsCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'insertMany';
        this._schema = {
            docs: 'fields',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.CREATE;
    }
}
exports.InsertDocumentsCommand = InsertDocumentsCommand;
class FindAndUpdateDocumentCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'findOneAndUpdate';
        this._schema = {
            filter: 'object',
            update: 'object',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.UPDATE;
    }
}
exports.FindAndUpdateDocumentCommand = FindAndUpdateDocumentCommand;
class UpdateDocumentCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'updateOne';
        this._schema = {
            filter: 'object',
            update: 'object',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.UPDATE;
    }
}
exports.UpdateDocumentCommand = UpdateDocumentCommand;
class UpdateDocumentsCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'updateMany';
        this._schema = {
            filter: 'object',
            update: 'object',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.UPDATE;
    }
}
exports.UpdateDocumentsCommand = UpdateDocumentsCommand;
class FindAndReplaceDocumentCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'findOneAndReplace';
        this._schema = {
            filter: 'object',
            replacement: 'object',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.UPDATE;
    }
}
exports.FindAndReplaceDocumentCommand = FindAndReplaceDocumentCommand;
class ReplaceDocumentCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'replaceOne';
        this._schema = {
            filter: 'object',
            doc: 'object',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.UPDATE;
    }
}
exports.ReplaceDocumentCommand = ReplaceDocumentCommand;
class FindAndDeleteDocumentCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'findOneAndDelete';
        this._schema = {
            filter: 'object',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.DELETE;
    }
}
exports.FindAndDeleteDocumentCommand = FindAndDeleteDocumentCommand;
class DeleteDocumentCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'deleteOne';
        this._schema = {
            filter: 'object',
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.DELETE;
    }
}
exports.DeleteDocumentCommand = DeleteDocumentCommand;
class DeleteDocumentsCommand extends BaseCommand {
    constructor() {
        super(...arguments);
        this.name = 'deleteMany';
        this._schema = {
            filter: {
                type: 'object',
                required: false,
            },
            options: {
                type: 'object',
                required: false,
            },
        };
        this._permission = constant_1.PERMISSION.DELETE;
    }
}
exports.DeleteDocumentsCommand = DeleteDocumentsCommand;
