// @alicloud/mp-db-service/dist/umd/mongo/model/query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const mpserverless_core_1 = require("@alicloud/mpserverless-core");
const codec_1 = require("../codec");
const collection_1 = require("./collection");
const command_1 = require("./command");
const utils_1 = require("../utils");
const constant_1 = require("../constant");
class Query {
    constructor() {
        this.encoder = new codec_1.JSONEncoder();
        this.decoder = new codec_1.JSONDecoder();
    }
    collection(name) {
        this.coll = new collection_1.Collection(name);
        return this;
    }
    aggregate(pipeline, options) {
        this.comm = new command_1.AggregateCommand({ pipeline, options });
        return this.execute();
    }
    count(query, options) {
        this.comm = new command_1.CountCommand({ query, options });
        return this.execute();
    }
    distinct(key, query, options) {
        this.comm = new command_1.DistinctCommand({ key, query, options });
        return this.execute();
    }
    findOne(query, options) {
        this.comm = new command_1.FindDocumentCommand({ query, options });
        return this.execute();
    }
    find(query, options) {
        this.comm = new command_1.FindDocumentsCommand({ query, options });
        return this.execute();
    }
    insertOne(doc, options) {
        this.comm = new command_1.InsertDocumentCommand({ doc, options });
        return this.execute();
    }
    insertMany(docs, options) {
        this.comm = new command_1.InsertDocumentsCommand({ docs, options });
        return this.execute();
    }
    findOneAndUpdate(filter, update, options) {
        this.comm = new command_1.FindAndUpdateDocumentCommand({ filter, update, options });
        return this.execute();
    }
    updateOne(filter, update, options) {
        this.comm = new command_1.UpdateDocumentCommand({ filter, update, options });
        return this.execute();
    }
    updateMany(filter, update, options) {
        this.comm = new command_1.UpdateDocumentsCommand({ filter, update, options });
        return this.execute();
    }
    findOneAndReplace(filter, replacement, options) {
        this.comm = new command_1.FindAndReplaceDocumentCommand({ filter, replacement, options });
        return this.execute();
    }
    replaceOne(filter, doc, options) {
        this.comm = new command_1.ReplaceDocumentCommand({ filter, doc, options });
        return this.execute();
    }
    findOneAndDelete(filter, options) {
        this.comm = new command_1.FindAndDeleteDocumentCommand({ filter, options });
        return this.execute();
    }
    deleteOne(filter, options) {
        this.comm = new command_1.DeleteDocumentCommand({ filter, options });
        return this.execute();
    }
    deleteMany(filter, options) {
        this.comm = new command_1.DeleteDocumentsCommand({ filter, options });
        return this.execute();
    }
    validate() {
        const v = new utils_1.MongoValidator();
        v.validate(this.comm.schema, this.encoder.encode(this.comm.argMap));
    }
    execute(..._) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            mpserverless_core_1.assert(this.comm, '[DBService]缺少 command 参数');
            this.validate();
            switch (this.comm.permission) {
                case constant_1.PERMISSION.AGGREGATE:
                case constant_1.PERMISSION.WRITE:
                case constant_1.PERMISSION.READ:
                    mpserverless_core_1.assert(this.collection, '[DBService]缺少 collection 参数');
                    break;
                default:
            }
            const query = Object.assign({}, this.coll && this.coll.inspect(), this.comm.inspect());
            return this.encoder.encode(query);
        });
    }
}
exports.Query = Query;
