// @alicloud/mp-db-service/dist/umd/mongo/utils/rules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ErrorMessages;
(function (ErrorMessages) {
    ErrorMessages["INVALID_TYPE"] = "field type is invalid";
    ErrorMessages["NOT_ARRAY"] = "field is not an array";
    ErrorMessages["ILLEGAL"] = "field should not contain illegal character";
})(ErrorMessages || (ErrorMessages = {}));
function ruleOfField(_, value) {
    function validateField(name) {
        if (/[\.\$]/.test(name)) {
            return ErrorMessages.ILLEGAL;
        }
    }
    if (Array.isArray(value)) {
        return undefined;
    }
    if (typeof value === 'string') {
        return validateField(value);
    }
    let result = null;
    for (const key of Object.keys(value)) {
        result = validateField(key);
        if (typeof result === 'string') {
            return result;
        }
        if (typeof value[key] === 'object') {
            return ruleOfField(_, value[key]);
        }
    }
    if (result === null) {
        return ErrorMessages.INVALID_TYPE;
    }
}
function ruleOfFields(_, value) {
    if (!Array.isArray(value)) {
        return ErrorMessages.NOT_ARRAY;
    }
    let err;
    let i = 0;
    while (!err && i < value.length) {
        err = ruleOfField(null, value[i]);
        i++;
    }
    return err;
}
const rules = {
    ruleOfField,
    ruleOfFields,
};
exports.rules = rules;
