// @alicloud/mp-db-service/dist/esm/mongo/utils/rules.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var ErrorMessages;
    (function (ErrorMessages) {
        ErrorMessages["INVALID_TYPE"] = "field type is invalid";
        ErrorMessages["NOT_ARRAY"] = "field is not an array";
        ErrorMessages["ILLEGAL"] = "field should not contain illegal character";
    })(ErrorMessages || (ErrorMessages = {}));
    function ruleOfField(_, value) {
        function validateField(name) {
            if (/[\.\$]/.test(name)) {
                return ErrorMessages.ILLEGAL;
            }
        }
        if (Array.isArray(value)) {
            return undefined;
        }
        if (typeof value === 'string') {
            return validateField(value);
        }
        var result = null;
        for (var _i = 0, _a = Object.keys(value); _i < _a.length; _i++) {
            var key = _a[_i];
            result = validateField(key);
            if (typeof result === 'string') {
                return result;
            }
            if (typeof value[key] === 'object') {
                return ruleOfField(_, value[key]);
            }
        }
        if (result === null) {
            return ErrorMessages.INVALID_TYPE;
        }
    }
    function ruleOfFields(_, value) {
        if (!Array.isArray(value)) {
            return ErrorMessages.NOT_ARRAY;
        }
        var err;
        var i = 0;
        while (!err && i < value.length) {
            err = ruleOfField(null, value[i]);
            i++;
        }
        return err;
    }
    var rules = {
        ruleOfField: ruleOfField,
        ruleOfFields: ruleOfFields,
    };
    exports.rules = rules;
});
