// @alicloud/mp-function-service/dist/esm/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@alicloud/mpserverless-core"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var mpserverless_core_1 = require("@alicloud/mpserverless-core");
    var FunctionService = (function (_super) {
        tslib_1.__extends(FunctionService, _super);
        function FunctionService() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FunctionService.prototype.invoke = function (functionTarget, functionArgs) {
            if (functionArgs === void 0) { functionArgs = {}; }
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                var encoder, response;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            encoder = this.getEncoder();
                            encoder.setBodyField({
                                method: 'serverless.function.runtime.invoke',
                                params: {
                                    functionTarget: functionTarget,
                                    functionArgs: functionArgs,
                                },
                            });
                            return [4, this.transport.request(encoder)];
                        case 1:
                            response = _a.sent();
                            return [2, tslib_1.__assign(tslib_1.__assign({}, response.body), { requestId: response.headers['request-id'] })];
                    }
                });
            });
        };
        return FunctionService;
    }(mpserverless_core_1.BaseService));
    exports.FunctionService = FunctionService;
});
