// @alicloud/mp-db-service/dist/esm/mongo/codec/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@alicloud/mpserverless-core"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var mpserverless_core_1 = require("@alicloud/mpserverless-core");
    var REGEXP_FLAGS = ['i', 'm', 'u', 'g'];
    function _isByProto(val, type) {
        return Object.prototype.toString.call(val) === "[object " + type + "]";
    }
    exports._isByProto = _isByProto;
    function isArray(val) {
        return Array.isArray(val);
    }
    exports.isArray = isArray;
    function isObject(val) {
        return val !== null && typeof val === 'object' && !isArray(val);
    }
    exports.isObject = isObject;
    function isString(val) {
        return typeof val === 'string' || _isByProto(val, 'String');
    }
    exports.isString = isString;
    var JSONEncoder = (function (_super) {
        tslib_1.__extends(JSONEncoder, _super);
        function JSONEncoder() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        JSONEncoder.prototype.encode = function (data) {
            var _this = this;
            if (data instanceof RegExp) {
                return this.toRegexp(data);
            }
            if (data instanceof Date) {
                return this.toDate(data);
            }
            if (isArray(data)) {
                return data.map(function (d) { return _this.encode(d); });
            }
            if (isObject(data)) {
                return Object.keys(data).reduce(function (accu, k) {
                    accu[k] = _this.encode(data[k]);
                    return accu;
                }, {});
            }
            return data;
        };
        JSONEncoder.prototype.toDate = function (val) {
            return val.toISOString();
        };
        JSONEncoder.prototype.toRegexp = function (val) {
            return "/" + val.source.replace(/\\\//g, '/') + "/" + val.flags;
        };
        return JSONEncoder;
    }(mpserverless_core_1.BaseEncoder));
    exports.JSONEncoder = JSONEncoder;
    var JSONDecoder = (function (_super) {
        tslib_1.__extends(JSONDecoder, _super);
        function JSONDecoder() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        JSONDecoder.prototype.decode = function (data) {
            var _this = this;
            if (this.isRegexp(data)) {
                return this.toRegexp(data);
            }
            if (this.isDate(data)) {
                return this.toDate(data);
            }
            if (isArray(data)) {
                return data.map(function (d) { return _this.decode(d); });
            }
            if (isObject(data)) {
                return Object.keys(data).reduce(function (accu, k) {
                    accu[k] = _this.decode(data[k]);
                    return accu;
                }, {});
            }
            return data;
        };
        JSONDecoder.prototype.isDate = function (val) {
            return isString(val) && /^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/.test(val) && !isNaN(Date.parse(val));
        };
        JSONDecoder.prototype.isRegexp = function (val) {
            if (isString(val)) {
                var areFlagsValid = true;
                var parts = val.split('/');
                var last = parts[parts.length - 1];
                if (last) {
                    areFlagsValid = last.split('').reduce(function (expr, p) {
                        return expr === true && REGEXP_FLAGS.indexOf(p) > -1;
                    }, true);
                }
                return parts.length >= 2 && areFlagsValid;
            }
            return false;
        };
        JSONDecoder.prototype.toDate = function (val) {
            return new Date(val);
        };
        JSONDecoder.prototype.toRegexp = function (val) {
            var firstIndex = val.indexOf('/');
            var lastIndex = val.lastIndexOf('/');
            return new RegExp(val.slice(firstIndex + 1, lastIndex), val.slice(lastIndex + 1));
        };
        return JSONDecoder;
    }(mpserverless_core_1.BaseDecoder));
    exports.JSONDecoder = JSONDecoder;
});
