// @alicloud/mp-db-service/dist/umd/mongo/codec/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mpserverless_core_1 = require("@alicloud/mpserverless-core");
const REGEXP_FLAGS = ['i', 'm', 'u', 'g'];
function _isByProto(val, type) {
    return Object.prototype.toString.call(val) === `[object ${type}]`;
}
exports._isByProto = _isByProto;
function isArray(val) {
    return Array.isArray(val);
}
exports.isArray = isArray;
function isObject(val) {
    return val !== null && typeof val === 'object' && !isArray(val);
}
exports.isObject = isObject;
function isString(val) {
    return typeof val === 'string' || _isByProto(val, 'String');
}
exports.isString = isString;
class JSONEncoder extends mpserverless_core_1.BaseEncoder {
    encode(data) {
        if (data instanceof RegExp) {
            return this.toRegexp(data);
        }
        if (data instanceof Date) {
            return this.toDate(data);
        }
        if (isArray(data)) {
            return data.map(d => this.encode(d));
        }
        if (isObject(data)) {
            return Object.keys(data).reduce((accu, k) => {
                accu[k] = this.encode(data[k]);
                return accu;
            }, {});
        }
        return data;
    }
    toDate(val) {
        return val.toISOString();
    }
    toRegexp(val) {
        return `/${val.source.replace(/\\\//g, '/')}/${val.flags}`;
    }
}
exports.JSONEncoder = JSONEncoder;
class JSONDecoder extends mpserverless_core_1.BaseDecoder {
    decode(data) {
        if (this.isRegexp(data)) {
            return this.toRegexp(data);
        }
        if (this.isDate(data)) {
            return this.toDate(data);
        }
        if (isArray(data)) {
            return data.map(d => this.decode(d));
        }
        if (isObject(data)) {
            return Object.keys(data).reduce((accu, k) => {
                accu[k] = this.decode(data[k]);
                return accu;
            }, {});
        }
        return data;
    }
    isDate(val) {
        return isString(val) && /^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/.test(val) && !isNaN(Date.parse(val));
    }
    isRegexp(val) {
        if (isString(val)) {
            let areFlagsValid = true;
            const parts = val.split('/');
            const last = parts[parts.length - 1];
            if (last) {
                areFlagsValid = last.split('').reduce((expr, p) => {
                    return expr === true && REGEXP_FLAGS.indexOf(p) > -1;
                }, true);
            }
            return parts.length >= 2 && areFlagsValid;
        }
        return false;
    }
    toDate(val) {
        return new Date(val);
    }
    toRegexp(val) {
        const firstIndex = val.indexOf('/');
        const lastIndex = val.lastIndexOf('/');
        return new RegExp(val.slice(firstIndex + 1, lastIndex), val.slice(lastIndex + 1));
    }
}
exports.JSONDecoder = JSONDecoder;
