// @alicloud/mp-db-service/dist/esm/mongo/model/query.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@alicloud/mpserverless-core", "../codec", "./collection", "./command", "../utils", "../constant"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var mpserverless_core_1 = require("@alicloud/mpserverless-core");
    var codec_1 = require("../codec");
    var collection_1 = require("./collection");
    var command_1 = require("./command");
    var utils_1 = require("../utils");
    var constant_1 = require("../constant");
    var Query = (function () {
        function Query() {
            this.encoder = new codec_1.JSONEncoder();
            this.decoder = new codec_1.JSONDecoder();
        }
        Query.prototype.collection = function (name) {
            this.coll = new collection_1.Collection(name);
            return this;
        };
        Query.prototype.aggregate = function (pipeline, options) {
            this.comm = new command_1.AggregateCommand({ pipeline: pipeline, options: options });
            return this.execute();
        };
        Query.prototype.count = function (query, options) {
            this.comm = new command_1.CountCommand({ query: query, options: options });
            return this.execute();
        };
        Query.prototype.distinct = function (key, query, options) {
            this.comm = new command_1.DistinctCommand({ key: key, query: query, options: options });
            return this.execute();
        };
        Query.prototype.findOne = function (query, options) {
            this.comm = new command_1.FindDocumentCommand({ query: query, options: options });
            return this.execute();
        };
        Query.prototype.find = function (query, options) {
            this.comm = new command_1.FindDocumentsCommand({ query: query, options: options });
            return this.execute();
        };
        Query.prototype.insertOne = function (doc, options) {
            this.comm = new command_1.InsertDocumentCommand({ doc: doc, options: options });
            return this.execute();
        };
        Query.prototype.insertMany = function (docs, options) {
            this.comm = new command_1.InsertDocumentsCommand({ docs: docs, options: options });
            return this.execute();
        };
        Query.prototype.findOneAndUpdate = function (filter, update, options) {
            this.comm = new command_1.FindAndUpdateDocumentCommand({ filter: filter, update: update, options: options });
            return this.execute();
        };
        Query.prototype.updateOne = function (filter, update, options) {
            this.comm = new command_1.UpdateDocumentCommand({ filter: filter, update: update, options: options });
            return this.execute();
        };
        Query.prototype.updateMany = function (filter, update, options) {
            this.comm = new command_1.UpdateDocumentsCommand({ filter: filter, update: update, options: options });
            return this.execute();
        };
        Query.prototype.findOneAndReplace = function (filter, replacement, options) {
            this.comm = new command_1.FindAndReplaceDocumentCommand({ filter: filter, replacement: replacement, options: options });
            return this.execute();
        };
        Query.prototype.replaceOne = function (filter, doc, options) {
            this.comm = new command_1.ReplaceDocumentCommand({ filter: filter, doc: doc, options: options });
            return this.execute();
        };
        Query.prototype.findOneAndDelete = function (filter, options) {
            this.comm = new command_1.FindAndDeleteDocumentCommand({ filter: filter, options: options });
            return this.execute();
        };
        Query.prototype.deleteOne = function (filter, options) {
            this.comm = new command_1.DeleteDocumentCommand({ filter: filter, options: options });
            return this.execute();
        };
        Query.prototype.deleteMany = function (filter, options) {
            this.comm = new command_1.DeleteDocumentsCommand({ filter: filter, options: options });
            return this.execute();
        };
        Query.prototype.validate = function () {
            var v = new utils_1.MongoValidator();
            v.validate(this.comm.schema, this.encoder.encode(this.comm.argMap));
        };
        Query.prototype.execute = function () {
            var _ = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                _[_i] = arguments[_i];
            }
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                var query;
                return tslib_1.__generator(this, function (_a) {
                    mpserverless_core_1.assert(this.comm, '[DBService]缺少 command 参数');
                    this.validate();
                    switch (this.comm.permission) {
                        case constant_1.PERMISSION.AGGREGATE:
                        case constant_1.PERMISSION.WRITE:
                        case constant_1.PERMISSION.READ:
                            mpserverless_core_1.assert(this.collection, '[DBService]缺少 collection 参数');
                            break;
                        default:
                    }
                    query = Object.assign({}, this.coll && this.coll.inspect(), this.comm.inspect());
                    return [2, this.encoder.encode(query)];
                });
            });
        };
        return Query;
    }());
    exports.Query = Query;
});
