// @alicloud/mp-db-service/dist/esm/index.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@alicloud/mpserverless-core", "./mongo"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var mpserverless_core_1 = require("@alicloud/mpserverless-core");
    var mongo_1 = require("./mongo");
    var QueryService = (function (_super) {
        tslib_1.__extends(QueryService, _super);
        function QueryService() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        QueryService.prototype.execute = function () {
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                var params, request, response;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4, _super.prototype.execute.call(this)];
                        case 1:
                            params = _a.sent();
                            request = this.getEncoder()
                                .setBodyField({
                                method: 'serverless.db.default.execute',
                                params: params,
                            });
                            return [4, this.getTransport().request(request)];
                        case 2:
                            response = _a.sent();
                            return [2, (new mongo_1.Result(response.body)).inspect()];
                    }
                });
            });
        };
        return QueryService;
    }(mongo_1.Query));
    exports.QueryService = QueryService;
    var DbService = (function (_super) {
        tslib_1.__extends(DbService, _super);
        function DbService() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        DbService.prototype.collection = function (name) {
            var _this = this;
            var q = (new QueryService()).collection(name);
            q.getEncoder = function () { return _this.getEncoder(); };
            q.getTransport = function () { return _this.transport; };
            return q;
        };
        return DbService;
    }(mpserverless_core_1.BaseService));
    exports.DbService = DbService;
});
